### PR TITLE
Adds ability to validate XML files against specified XSD files

### PR DESCRIPTION
**Pull request by fhaScireum**
- Fixes: SIRI-255

### CODE REVIEW

File: src/main/java/sirius/biz/jobs/batch/file/XMLImportJobFactory.java [hunk 1]
```diff
@@ -8,13 +8,27 @@
...
  */
 public abstract class XMLImportJobFactory extends FileImportJobFactory {
 
+    protected final BooleanParameter requireValidFile =
+            new BooleanParameter("requireValidFile", "$XMLImportJobFactory.requireValidFile").hidden()
+                                                                                                       .logInSystem();
+
+    @Override
+    protected void collectParameters(Consumer<Parameter<?, ?>> parameterCollector) {
+        parameterCollector.accept(requireValidFile);
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/batch/file/XMLImportJobFactory.java:28]: only add if getXsdResourcePath() != null?

File: src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import javax.xml.XMLConstants;
+import javax.xml.transform.Source;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+import java.io.IOException;
+import java.util.Locale;
+
+/**
+ * wra
```
Comment (bwiedmann) [inline src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:30]: Add javadoc

File: src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ */
+public class SimpleXMLValidator {
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:32]: Why is this simple and what would be a complex one?

File: src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java (hunk 2 above)
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:30]: :)

**Review by andyHa — CHANGES_REQUESTED** [review]
...

File: src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java (hunk 2 above)
Comment (fhaScireum) [reply src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:30]: :)

File: src/main/resources/biz_de.properties [hunk 4]
```diff
@@ -821,3 +821,6 @@ VirtualObject.rename = Umbenennen
 VirtualObject.replaceContents = Inhalt ersetzen
 VirtualObject.unreference = Referenzierung löschen
 VirtualObject.versions = Versionen
+XMLImportJob.importCanceled = Der Import der Datei '${fileName}' wurde verhindert, da diese invalide ist.
```
Comment (bwiedmann) [inline src/main/resources/biz_de.properties:824]: Du willst damit ja ausdrücken dass diese nicht gegen das XSD Schema validierte oder? Gerne noch etwas sprechender formulieren 👍

**Review by bwiedmann — CHANGES_REQUESTED** [review]
...

File: src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java (hunk 3 above)
Comment (fhaScireum) [reply src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:32]: I would say that it is simple as only one xml and xsd source can be provided without additional settings. But I can rename it to XMLValidator as this class might evolve in time.

Comment (idlira) [reply src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:32]: XsdParser or XsdValidator would be my call

Comment (fhaScireum) [reply src/main/java/sirius/biz/jobs/batch/file/SimpleXMLValidator.java:32]: In my opinion, XsdValidator makes it seem like it validates XSD's.

File: src/main/resources/biz_de.properties [hunk 5]
```diff
@@ -821,6 +821,6 @@ VirtualObject.rename = Umbenennen
 VirtualObject.replaceContents = Inhalt ersetzen
 VirtualObject.unreference = Referenzierung löschen
 VirtualObject.versions = Versionen
-XMLImportJob.importCanceled = Der Import der Datei '${fileName}' wurde verhindert, da diese invalide ist.
+XMLImportJob.importCanceled = Der Import der Datei '${fileName}' wurde verhindert, da diese gegenüber dem hinterlegten XML Schema ungültig ist.
```
Comment (idlira) [inline src/main/resources/biz_de.properties:824]:
```suggestion
XMLImportJob.importCanceled = Der Import der Datei '${fileName}' wurde abgebrochen, da diese gegenüber dem vordefinierte XML Schema ungültig ist.
```

File: src/main/resources/biz_en.properties [hunk 6]
```diff
@@ -777,6 +777,6 @@ VirtualObject.rename = Rename
 VirtualObject.replaceContents = Replace contents
 VirtualObject.unreference = Delete reference
 VirtualObject.versions = Versions
-XMLImportJob.importCanceled = Importing the file '${fileName}' was prevented as the file is invalid.
+XMLImportJob.importCanceled = Importing the file '${fileName}' was prevented as the file is invalid against the stored XML Schema.
```
Comment (idlira) [inline src/main/resources/biz_en.properties:780]:
```suggestion
XMLImportJob.importCanceled = Importing the file '${fileName}' was cancelled as the file is invalid against the defined XML Schema.
```

**Review by idlira — CHANGES_REQUESTED** [review]
Few suggestions.
What export job, in another PR?

**Comment by fhaScireum** [comment]
> Few suggestions.
> What export job, in another PR?

Yes

**Review by idlira — APPROVED** [review]

**Review by bwiedmann — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by bwiedmann — APPROVED** [review]

File: src/main/java/sirius/biz/jobs/batch/file/XMLValidator.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        validator.setProperty("http://apache.org/xml/properties/locale", locale);
+
+        XMLValidatorErrorHandler errorHandler = new XMLValidatorErrorHandler(process);
+        validator.setErrorHandler(errorHandler);
+
+        try {
+            validator.validate(xml);
+        } catch (SAXParseException | IOException e) {
+            // Cause of the exception is already logged via validator's error handler.
+            Exceptions.ignore(e);
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/jobs/batch/file/XMLValidator.java:71]: Error handler will only handle SAXParseException. So we will swallow IOExceptions here.

File: src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java [hunk 8]
```diff
@@ -27,6 +47,93 @@
      */
     protected XMLImportJob(XMLImportJobFactory factory, ProcessContext process) {
         super(factory.fileParameter, process);
+        requireValidFile = process.getParameter(factory.requireValidFile).orElse(false);
+        xsdResourcePath = factory.getXsdResourcePath();
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java:51]: Is therre a use case for having a validation on xml file import and ignoring the result? If not, then why not just use the xsdPath parameter to determine if a validation should happen?

**Review by mkeckmkeck — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java (hunk 8 above)
Comment (idlira) [reply src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java:51]: The path dictates if a validation should happen. The flag controls the behavior of the job:
- Import: Abort or try to import anyway.
- Export: finish the job with either failed or success.
In both cases, validation errors will be logged accordingly as warning/error.

Comment (mkeckmkeck) [reply src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java:51]: We are inside an import job here. If you have a use case for validation and ignoring the results, than that is fine. Just I don't see that use case :-)

Comment (mkeckmkeck) [reply src/main/java/sirius/biz/jobs/batch/file/XMLImportJob.java:51]: Fine for me

**Review by bwiedmann — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**